dlog/commitment/src/commitment.rs

```diff
@@ -160,57 +179,176 @@ impl<G: AffineCurve> SRS<G>
+                p += &(p_i.scale(scale));
+                // p_comm += &(comm_i.mul(scale));
+                scale *= &polyscale;
+
+                match degree_bound {
+                    Some(m) => {
+                        // mixing in the shifted polynom since degree is bounded
+                        p += &(p_i.shiftr(self.g.len() - m).scale(scale));
+                        // p_comm += &(shifted_comm_i.mul(scale));
+                        scale *= &polyscale;
+                    }
+                    _ => {}
                 }
-                _ => {}
             }
-            scale *= &polyscale;
-            // always mixing in the unshifted polynom
-            p += &(x.0.scale(scale));
-        }
+            p
+        };
 
-        // randomise/scale the eval powers
-        let mut scale = Fr::<G>::one();
-        let b = elm.iter().map
-        (
-            |elm|
-            {
-                let mut acc = Fr::<G>::one();
-                DensePolynomial::<Fr<G>>::from_coefficients_vec((0..d).map(|_| {let r = acc; acc *= &elm; r}).collect::<Vec<_>>())
+        // TODO: Add blindings to the commitments. Opening will require knowing the blinding factor
+        // for each commitment in the batch.
+        let blinding_factor = Fr::<G>::zero();
+
+        let padded_length = 1 << rounds;
+
+        // b_j = sum_i r^i elm_i^j
+        let b_init = {
+            // randomise/scale the eval powers
+            let mut scale = Fr::<G>::one();
+            let mut res: Vec<Fr<G>> = (0..padded_length).map(|_| Fr::<G>::zero()).collect();
+            for e in elm {
+                for (i, t) in pows(padded_length, *e).iter().enumerate() {
+                    res[i] += &(scale * t);
+                }
+                scale *= &evalscale;
             }
-        ).fold(DensePolynomial::<Fr<G>>::zero(), |x, y| {scale *= &evalscale; &x + &y.scale(scale)});
-        
+            res
+        };
+
+        let padding = padded_length - self.g.len();
+        let mut g = self.g.clone();
+        g.extend(vec![G::zero(); padding]);
+
+        let mut a = p.coeffs;
+        a.extend(vec![Fr::<G>::zero(); padded_length - a.len()]);
+
+        let mut b = b_init.clone();
 
         let mut sponge = EFqSponge::new(oracle_params.clone());
-        match Self::recursion (&self, &p, &b, &mut sponge)
-        {
-            Ok(proof) => Ok(OpeningProof::<G>{lr: proof.0, s: proof.1}),
-            Err(err) => Err(err)
+
+        let mut lr = vec![];
+
+        let mut blinders = vec![];
+
+        let mut chals = vec![];
+        let mut chal_invs = vec![];
+
+        for _ in 0..rounds {
+            let n = g.len() / 2;
+            let (g_lo, g_hi) = (g[0..n].to_vec(), g[n..].to_vec());
+            let (a_lo, a_hi) = (&a[0..n], &a[n..]);
+            let (b_lo, b_hi) = (&b[0..n], &b[n..]);
+
+            let rand_l = Fr::<G>::rand(rng);
+            let rand_r = Fr::<G>::rand(rng);
+
+            let l = {
```

Comment:
I would use the following code to compute L & R:

            let l = {
                let a_lo_b_hi = inner_prod(a_lo, b_hi);
                let a_lo: Vec<_> = a_lo.iter().map(|x| x.into_repr()).collect();
                (VariableBaseMSM::multi_scalar_mul(&g_hi, &a_lo)
                    + &(self.h.mul(rand_l) + &u.mul(a_lo_b_hi)))
                    .into_affine()
            };

            let r = {
                let a_hi_b_lo = inner_prod(a_hi, b_lo);
                let a_hi: Vec<_> = a_hi.iter().map(|x| x.into_repr()).collect();
                (VariableBaseMSM::multi_scalar_mul(&g_lo, &a_hi)
                    + &(self.h.mul(rand_r) + &u.mul(a_hi_b_lo)))
                    .into_affine()
            };